src/interfaces/swig/shogun.i

```diff
@@ -34,16 +34,357 @@
 #ifdef SWIGPYTHON
 #include <object.h>
 %{
-    static int print_sgobject(PyObject *pyobj, FILE *f, int flags);
+	static int print_sgobject(PyObject *pyobj, FILE *f, int flags);
+
+	#include <shogun/lib/any.h>
+	#include <shogun/io/SGIO.h>
+	#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+	extern "C" {
+		#include <numpy/arrayobject.h>
+	}
+	#define SG_TO_NUMPY_TYPE_STRUCT(SG_TYPE, NPY_TYPE) \
+	template <>                                        \
+	struct sg_to_npy_type<SG_TYPE>                     \
+	{                                                  \
+		const static NPY_TYPES type = NPY_TYPE;        \
+	};
+
+	namespace shogun
+	{
+		template <typename T>
+		struct sg_to_npy_type {};
+		SG_TO_NUMPY_TYPE_STRUCT(bool,          NPY_BOOL)
+#ifdef PYTHON3 // str -> unicode for python3
+		SG_TO_NUMPY_TYPE_STRUCT(char,          NPY_UNICODE)
+#else
+		SG_TO_NUMPY_TYPE_STRUCT(char,          NPY_STRING)
+#endif
+		SG_TO_NUMPY_TYPE_STRUCT(int8_t,        NPY_INT8)
+		SG_TO_NUMPY_TYPE_STRUCT(uint8_t,       NPY_UINT8)
+		SG_TO_NUMPY_TYPE_STRUCT(int16_t,       NPY_INT16)
+		SG_TO_NUMPY_TYPE_STRUCT(uint16_t,      NPY_UINT16)
+		SG_TO_NUMPY_TYPE_STRUCT(int32_t,       NPY_INT32)
+		SG_TO_NUMPY_TYPE_STRUCT(uint32_t,      NPY_UINT32)
+		SG_TO_NUMPY_TYPE_STRUCT(int64_t,       NPY_INT64)
+		SG_TO_NUMPY_TYPE_STRUCT(uint64_t,      NPY_UINT64)
+		SG_TO_NUMPY_TYPE_STRUCT(float32_t,     NPY_FLOAT32)
+		SG_TO_NUMPY_TYPE_STRUCT(float64_t,     NPY_FLOAT64)
+		SG_TO_NUMPY_TYPE_STRUCT(complex128_t,  NPY_CDOUBLE)
+		SG_TO_NUMPY_TYPE_STRUCT(floatmax_t,    NPY_LONGDOUBLE)
+		SG_TO_NUMPY_TYPE_STRUCT(PyObject*,     NPY_OBJECT)
+		SG_TO_NUMPY_TYPE_STRUCT(CSGObject*,    NPY_OBJECT)
+#undef SG_TO_NUMPY_TYPE_STRUCT
+
+		class PythonVisitor : public AnyVisitor {
+
+		public:
+			PythonVisitor(PyObject*& obj) : AnyVisitor(), m_py_obj(&obj) {}
+
+			void on(bool *v) final
+			{
+				handle_sg(v);
+			}
+
+			void on(int8_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(int16_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(int32_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(int64_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(float32_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(float64_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(floatmax_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(std::string *v) final
+			{
+			  handle_sg(v->c_str());
+			}
+
+			void on(CSGObject **v) final
+			{
+				handle_sg(v);
+			}
+
+			void on(char *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(uint8_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(uint16_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(uint32_t *v) final
+			{
+			  handle_sg(v);
+			}
+
+			void on(uint64_t *v) final
+			{
+				handle_sg(v);
+			}
+
+			void on(complex128_t *v) final
```

Comment:
Would it make sense to make all these on's less verbose?

Reply:
less verbose in what sense? compress it to one line?

Reply:
Not needing to define all overloads this explicitly.

Reply:
not sure you can because those are virtual functions? i havent tried, but i don't think you can put it all in a templated function..